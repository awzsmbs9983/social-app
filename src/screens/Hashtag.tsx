import React from 'react'
import {ListRenderItemInfo, Pressable} from 'react-native'
import {PostView} from '@atproto/api/dist/client/types/app/bsky/feed/defs'
import {msg} from '@lingui/macro'
import {useLingui} from '@lingui/react'
import {useFocusEffect} from '@react-navigation/native'
import {NativeStackScreenProps} from '@react-navigation/native-stack'

import {HITSLOP_10} from 'lib/constants'
import {useInitialNumToRender} from 'lib/hooks/useInitialNumToRender'
import {CommonNavigatorParams} from 'lib/routes/types'
import {shareUrl} from 'lib/sharing'
import {cleanError} from 'lib/strings/errors'
import {sanitizeHandle} from 'lib/strings/handles'
import {enforceLen} from 'lib/strings/helpers'
import {isNative} from 'platform/detection'
import {useSearchPostsQuery} from 'state/queries/search-posts'
import {useSetMinimalShellMode} from 'state/shell'
import {Post} from 'view/com/post/Post'
import {List} from 'view/com/util/List'
import {ViewHeader} from 'view/com/util/ViewHeader'
import {ArrowOutOfBox_Stroke2_Corner0_Rounded} from '#/components/icons/ArrowOutOfBox'
import {
  ListFooter,
  ListHeaderDesktop,
  ListMaybePlaceholder,
} from '#/components/Lists'

const renderItem = ({item}: ListRenderItemInfo<PostView>) => {
  return <Post post={item} />
}

const keyExtractor = (item: PostView, index: number) => {
  return `${item.uri}-${index}`
}

export default function HashtagScreen({
  route,
}: NativeStackScreenProps<CommonNavigatorParams, 'Hashtag'>) {
  const {tag, author} = route.params
  const setMinimalShellMode = useSetMinimalShellMode()
  const {_} = useLingui()
  const initialNumToRender = useInitialNumToRender()
  const [isPTR, setIsPTR] = React.useState(false)

  const fullTag = React.useMemo(() => {
    return `#${decodeURIComponent(tag)}`
  }, [tag])

  const queryParam = React.useMemo(() => {
    if (!author) return fullTag
    return `${fullTag} from:${sanitizeHandle(author)}`
  }, [fullTag, author])

  const headerTitle = React.useMemo(() => {
    return enforceLen(fullTag.toLowerCase(), 24, true, 'middle')
  }, [fullTag])

  const sanitizedAuthor = React.useMemo(() => {
    if (!author) return
    return sanitizeHandle(author)
  }, [author])

  const {
    data,
    isFetchingNextPage,
    isLoading,
    isError,
    error,
    refetch,
    fetchNextPage,
    hasNextPage,
  } = useSearchPostsQuery({query: queryParam})

  const posts = React.useMemo(() => {
    return data?.pages.flatMap(page => page.posts) || []
  }, [data])

  useFocusEffect(
    React.useCallback(() => {
      setMinimalShellMode(false)
    }, [setMinimalShellMode]),
  )

  const onShare = React.useCallback(() => {
    const url = new URL('https://bsky.app')
    url.pathname = `/hashtag/${decodeURIComponent(tag)}`
    if (author) {
      url.searchParams.set('author', author)
    }
    shareUrl(url.toString())
  }, [tag, author])

  const onRefresh = React.useCallback(async () => {
    setIsPTR(true)
    await refetch()
    setIsPTR(false)
  }, [refetch])

  const onEndReached = React.useCallback(() => {
    if (isFetchingNextPage || !hasNextPage || error) return
    fetchNextPage()
  }, [isFetchingNextPage, hasNextPage, error, fetchNextPage])

  return (
    <>
      <ViewHeader
        title={headerTitle}
        subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
        canGoBack
        renderButton={
          isNative
            ? () => (
                <Pressable
                  accessibilityRole="button"
                  onPress={onShare}
                  hitSlop={HITSLOP_10}>
                  <ArrowOutOfBox_Stroke2_Corner0_Rounded
                    size="lg"
                    onPress={onShare}
                  />
                </Pressable>
              )
            : undefined
        }
      />
      {posts.length < 1 ? (
        <ListMaybePlaceholder
          isLoading={isLoading}
          isError={isError}
          onRetry={refetch}
          emptyType="results"
          emptyMessage={_(msg`We couldn't find any results for that hashtag.`)}
        />
      ) : (
        <List
          data={posts}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          refreshing={isPTR}
          onRefresh={onRefresh}
          onEndReached={onEndReached}
          onEndReachedThreshold={4}
          // @ts-ignore web only -prf
          desktopFixedHeight
          ListHeaderComponent={
            <ListHeaderDesktop
              title={headerTitle}
              subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
            />
          }
          ListFooterComponent={
            <ListFooter
              isFetchingNextPage={isFetchingNextPage}
              error={cleanError(error)}
              onRetry={fetchNextPage}
            />
          }
          initialNumToRender={initialNumToRender}
          windowSize={11}
        />
      )}
    </>
  )
}
