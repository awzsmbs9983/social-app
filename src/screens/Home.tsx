import React from 'react'
import {Text, Button, View, SafeAreaView} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Home = ({navigation}: PrimaryTabScreenProps<'Home'>) => {
  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={{flex: 1}}>
        <Text>Hello world</Text>
        <Button
          title="Go to Jane's profile"
          onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
        />
      </View>
    </SafeAreaView>
  )
}
