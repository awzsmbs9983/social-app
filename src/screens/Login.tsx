import React from 'react'
import {Text, Button, View} from 'react-native'
import {Shell} from '../platform/shell'
import type {RootTabsScreenProps} from '../routes/types'
import {useStores} from '../state'

export function Login({navigation}: RootTabsScreenProps<'Login'>) {
  const store = useStores()
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Sign In</Text>
        <Button title="Login" onPress={() => store.session.setAuthed(true)} />
        <Button title="Sign Up" onPress={() => navigation.navigate('Signup')} />
      </View>
    </Shell>
  )
}
