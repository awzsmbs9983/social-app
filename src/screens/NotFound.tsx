import React from 'react'
import {Shell} from '../platform/shell'
import {Text, Button, View} from 'react-native'
import type {RootTabsScreenProps} from '../routes/types'

export const NotFound = ({navigation}: RootTabsScreenProps<'NotFound'>) => {
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Page not found</Text>
        <Button title="Home" onPress={() => navigation.navigate('Home')} />
      </View>
    </Shell>
  )
}
