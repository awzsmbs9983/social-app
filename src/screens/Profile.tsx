import React from 'react'
import {Text} from 'react-native'
import type {RootStackScreenProps} from '../routes/types'

export const Profile = ({route}: RootStackScreenProps<'Profile'>) => {
  return <Text>This is {route.params.name}'s profile</Text>
}
