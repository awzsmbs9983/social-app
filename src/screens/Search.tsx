import React from 'react'
import {SafeAreaView, ScrollView, Text, Button, View} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Search = ({navigation}: PrimaryTabScreenProps<'Search'>) => {
  return (
    <SafeAreaView>
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <View>
          <Text>Hello world</Text>
          <Button
            title="Go to Jane's profile"
            onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  )
}
