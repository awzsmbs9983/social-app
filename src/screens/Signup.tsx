import React from 'react'
import {Text, Button, View, ActivityIndicator} from 'react-native'
import {observer} from 'mobx-react-lite'
import {Shell} from '../platform/shell'
import type {RootTabsScreenProps} from '../routes/types'
import {useStores} from '../state'

export const Signup = observer(
  ({navigation}: RootTabsScreenProps<'Signup'>) => {
    const store = useStores()
    return (
      <Shell>
        <View style={{justifyContent: 'center', alignItems: 'center'}}>
          <Text style={{fontSize: 20, fontWeight: 'bold'}}>Create Account</Text>
          {store.session.uiError ?? <Text>{store.session.uiError}</Text>}
          {store.session.uiState === 'idle' ? (
            <>
              <Button
                title="Create new account"
                onPress={() =>
                  store.session.createTestAccount('http://localhost:1986')
                }
              />
              <Button
                title="Log in to an existing account"
                onPress={() => navigation.navigate('Login')}
              />
            </>
          ) : (
            <ActivityIndicator />
          )}
        </View>
      </Shell>
    )
  },
)
