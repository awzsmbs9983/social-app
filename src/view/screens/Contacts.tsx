import React, {useEffect, useState, useRef} from 'react'
import {StyleSheet, Text, TextInput, View} from 'react-native'
import {useSharedValue} from 'react-native-reanimated'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {ProfileFollows as ProfileFollowsComponent} from '../com/profile/ProfileFollows'
import {Selector} from '../com/util/Selector'
import {colors} from '../lib/styles'
import {ScreenParams} from '../routes'
import {useStores} from '../../state'

export const Contacts = ({navIdx, visible, params}: ScreenParams) => {
  const store = useStores()
  const selectorInterp = useSharedValue(0)

  useEffect(() => {
    if (visible) {
      store.nav.setTitle(navIdx, `Contacts`)
    }
  }, [store, visible])

  const [searchText, onChangeSearchText] = useState('')
  const inputRef = useRef<TextInput | null>(null)

  return (
    <View>
      <View style={styles.section}>
        <Text style={styles.title}>Contacts</Text>
      </View>
      <View style={styles.section}>
        <View style={styles.searchContainer}>
          <FontAwesomeIcon
            icon="magnifying-glass"
            size={16}
            style={styles.searchIcon}
          />
          <TextInput
            ref={inputRef}
            value={searchText}
            style={styles.searchInput}
            placeholder="Search"
            onChangeText={onChangeSearchText}
          />
        </View>
      </View>
      <Selector
        items={['All', 'Following', 'Scenes']}
        selectedIndex={0}
        swipeGestureInterp={selectorInterp}
      />
      {!!store.me.handle && <ProfileFollowsComponent name={store.me.handle} />}
    </View>
  )
}

const styles = StyleSheet.create({
  section: {
    backgroundColor: colors.white,
  },
  title: {
    fontSize: 30,
    fontWeight: 'bold',
    paddingHorizontal: 12,
    paddingVertical: 6,
  },

  searchContainer: {
    flexDirection: 'row',
    backgroundColor: colors.gray1,
    paddingHorizontal: 8,
    paddingVertical: 8,
    marginHorizontal: 10,
    marginBottom: 6,
    borderRadius: 4,
  },
  searchIcon: {
    color: colors.gray5,
    marginRight: 8,
  },
  searchInput: {
    flex: 1,
  },
})
