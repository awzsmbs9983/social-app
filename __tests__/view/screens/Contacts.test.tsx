import React from 'react'
import {Contacts} from '../../../src/view/screens/Contacts'
import renderer from 'react-test-renderer'
// import {render} from '../../../../jest/test-utils'

describe('Contacts', () => {
  const mockedProps = {
    navIdx: [0, 0] as [number, number],
    params: {},
    visible: true,
  }
  it('renders correctly', () => {
    const tree = renderer.create(<Contacts {...mockedProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
