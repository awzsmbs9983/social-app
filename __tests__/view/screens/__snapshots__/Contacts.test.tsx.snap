// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
          },
          Object {
            "fontSize": 30,
            "fontWeight": "bold",
            "paddingHorizontal": 12,
            "paddingVertical": 6,
          },
        ]
      }
    >
      Contacts
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f8f3f3",
          "borderRadius": 4,
          "flexDirection": "row",
          "marginBottom": 6,
          "marginHorizontal": 10,
          "paddingHorizontal": 8,
          "paddingVertical": 8,
        }
      }
    >
      <
        icon="magnifying-glass"
        size={16}
        style={
          Object {
            "color": "#645454",
            "marginRight": 8,
          }
        }
      />
      <TextInput
        onChangeText={[Function]}
        placeholder="Search"
        placeholderTextColor="#968d8d"
        style={
          Object {
            "color": "#000000",
            "flex": 1,
          }
        }
        value=""
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "paddingBottom": 12,
          "paddingHorizontal": 14,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#000000",
          "bottom": 0,
          "height": 4,
          "left": 0,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginRight": 14,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
            },
            Object {
              "color": "#000000",
              "fontSize": 16,
              "fontWeight": "600",
            },
          ]
        }
      >
        All
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginRight": 14,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
            },
            Object {
              "color": "#645454",
              "fontSize": 16,
              "fontWeight": "600",
            },
          ]
        }
      >
        Following
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginRight": 14,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
            },
            Object {
              "color": "#645454",
              "fontSize": 16,
              "fontWeight": "600",
            },
          ]
        }
      >
        Scenes
      </Text>
    </View>
  </View>
</View>
`;
